### PR TITLE
 fixing conflict error while start experiment

### PR DESCRIPTION
**Pull request by YaningX**
first 2 steps for fixing conflict error while start experiment, check cloudservice available by azure api, and async try to get another available vm if no available

### CODE REVIEW

File: open-hackathon-server/src/hackathon/hack/host_server_manager.py [hunk 1]
```diff
@@ -90,14 +90,25 @@ def get_available_docker_host(self, req_count, hackathon):
...
+                    if deployment.locked:
+                        available = False
+                        break;
+                if available:
+                    return docker_host
         if not self.util.is_local():
-            self.create_docker_host_vm(hackathon.id)
-        return None
+            self.create_docker_host_vm(hackathon_id)
+        self.scheduler.add_once("docker_host_manager", "get_available_docker_host", context=ctx, seconds=10)
```
Comment (juniwang) [inline open-hackathon-server/src/hackathon/hack/host_server_manager.py]: what's the purpose of this function?

File: open-hackathon-server/src/hackathon/docker/hosted_docker.py [hunk 2]
```diff
@@ -188,7 +188,12 @@ def start(self, unit, **kwargs):
...
         container_name = unit.get_name()
-        host_server = self.docker_host_manager.get_available_docker_host(1, hackathon)
+        ctx = Context(
+            req_count=1,
+            hackathon_id=hackathon.id,
+            azure_key_id=self.load_azure_key_id(experiment.id)
+        )
+        host_server = self.docker_host_manager.get_available_docker_host(ctx)
         if not host_server:
             return None
```
Comment (juniwang) [inline open-hackathon-server/src/hackathon/docker/hosted_docker.py]: You can combine the above 3 lines into one: `return self.docker_host_manager.get_available_docker_host(ctx)`

File: open-hackathon-server/src/hackathon/hack/host_server_manager.py [hunk 3]
```diff
@@ -90,14 +90,25 @@ def get_available_docker_host(self, req_count, hackathon):
         # it's more reasonable to launch VM when the existed ones are almost used up.
         # The new-created VM must run 'cloudvm service by default(either cloud-init or python remote ssh)
         # todo the VM public/private IP will change after reboot, need sync the IP in db with azure in this case
+        service = Service(azure_key_id=azure_key_id)
         for docker_host in vms:
             if self.hosted_docker.ping(docker_host):
-                return docker_host
+                service_name = docker_host.public_dns.split(".")[0]
+                deployments = service.get_hosted_service_properties(service_name, detail=True).deployments
```
Comment (juniwang) [inline open-hackathon-server/src/hackathon/hack/host_server_manager.py]: One more thing, the code should work on local environment as well where `local` means our local development environment and it doesn't connect to azure at all.

**Comment by YaningX** [comment]
Step 1 for fixing conflict error while start experiment.

**Pull request merged**